pkg/storegateway/indexheader/index/postings.go

```diff
@@ -298,130 +291,86 @@ type postingOffset struct {
-			// Posting format entry is as follows:
-			// │ ┌────────────────────────────────────────┐ │
-			// │ │  n = 2 <1b>                            │ │
-			// │ ├──────────────────────┬─────────────────┤ │
-			// │ │ len(name) <uvarint>  │ name <bytes>    │ │
-			// │ ├──────────────────────┼─────────────────┤ │
-			// │ │ len(value) <uvarint> │ value <bytes>   │ │
-			// │ ├──────────────────────┴─────────────────┤ │
-			// │ │  offset <uvarint64>                    │ │
-			// │ └────────────────────────────────────────┘ │
-			// First, let's skip n and name.
-			skipNAndName(&d, &buf)
-			value := d.UvarintStr() // Label value.
-			postingOffset := int64(d.Uvarint64())
-
-			if len(newSameRngs) > 0 {
-				// We added some ranges in previous iteration. Use next posting offset as end of all our new ranges.
-				for j := range newSameRngs {
-					newSameRngs[j].End = postingOffset - crc32.Size
-				}
-				rngs = append(rngs, newSameRngs...)
-				newSameRngs = newSameRngs[:0]
-			}
+	if i > 0 && e.offsets[i].value != value {
+		// Need to look from previous entry.
+		i--
+	}
 
-			for value >= wantedValue {
-				// If wantedValue is equals of greater than current value, loop over all given wanted values in the values until
-				// this is no longer true or there are no more values wanted.
-				// This ensures we cover case when someone asks for postingsOffset(name, value1, value1, value1).
-
-				// Record on the way if wanted value is equal to the current value.
-				if value == wantedValue {
-					newSameRngs = append(newSameRngs, index.Range{Start: postingOffset + postingLengthFieldSize})
-				}
-				valueIndex++
-				if valueIndex == len(values) {
-					break
-				}
-				wantedValue = values[valueIndex]
-			}
+	d.ResetAt(e.offsets[i].tableOff)
+	nAndNameSize := 0
+
+	for d.Err() == nil {
+		// Posting format entry is as follows:
+		// │ ┌────────────────────────────────────────┐ │
+		// │ │  n = 2 <1b>                            │ │
+		// │ ├──────────────────────┬─────────────────┤ │
+		// │ │ len(name) <uvarint>  │ name <bytes>    │ │
+		// │ ├──────────────────────┼─────────────────┤ │
+		// │ │ len(value) <uvarint> │ value <bytes>   │ │
+		// │ ├──────────────────────┴─────────────────┤ │
+		// │ │  offset <uvarint64>                    │ │
+		// │ └────────────────────────────────────────┘ │
+		skipNAndName(&d, &nAndNameSize)
+		currentValue := d.UvarintStr()
+		postingOffset := int64(d.Uvarint64())
+
+		if currentValue > value {
+			// There is no entry for value.
+			return index.Range{}, false, nil
+		}
+
+		if currentValue == value {
+			rng := index.Range{Start: postingOffset + postingLengthFieldSize}
 
 			if i+1 == len(e.offsets) {
 				// No more offsets for this name.
-				// Break this loop and record lastOffset on the way for ranges we just added if any.
-				for j := range newSameRngs {
-					newSameRngs[j].End = e.lastValOffset
-				}
-				rngs = append(rngs, newSameRngs...)
-				break
+				rng.End = e.lastValOffset
```

Comment:
[unrelated from this PR, non blocking] Given `lastValOffset` is used as the end range in case this is the last value for the given label name, I'm wondering if we could find a better name for `lastValOffset`, cause it sounds like "offset of the start of the last value" and not "offset of the end of the last value". Makes sense or am I missing something?